app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp

```diff
@@ -1,214 +1,702 @@
+                case AndroidPixelFormat::RGBX8888:
+                    format = gpu::format::RGBA8888Unorm;
+                    break;
+
+                case AndroidPixelFormat::RGB565:
+                    format = gpu::format::RGB565Unorm;
                     break;
+
+                default:
+                    throw exception("Unknown format in buffer: '{}' ({})", ToString(graphicBuffer.format), static_cast<u32>(graphicBuffer.format));
+            }
+
+            auto &handle{graphicBuffer.graphicHandle};
+            if (handle.magic != NvGraphicHandle::Magic)
+                throw exception("Unexpected NvGraphicHandle magic: {}", handle.surfaceCount);
+            else if (handle.surfaceCount < 1)
+                throw exception("At least one surface is required in a buffer: {}", handle.surfaceCount);
+            else if (handle.surfaceCount > 1)
+                throw exception("Multi-planar surfaces are not supported: {}", handle.surfaceCount);
+
+            auto &surface{graphicBuffer.graphicHandle.surfaces.at(0)};
+            if (surface.scanFormat != NvDisplayScanFormat::Progressive)
+                throw exception("Non-Progressive surfaces are not supported: {}", ToString(surface.scanFormat));
+
+            std::shared_ptr<nvdrv::device::NvMap::NvMapObject> nvBuffer{};
+            {
+                auto driver{nvdrv::driver.lock()};
+                auto nvmap{driver->nvMap.lock()};
+                if (surface.nvmapHandle) {
+                    nvBuffer = nvmap->GetObject(surface.nvmapHandle);
+                } else {
+                    std::shared_lock nvmapLock(nvmap->mapMutex);
+                    for (const auto &object : nvmap->maps) {
+                        if (object->id == handle.nvmapId) {
+                            nvBuffer = object;
+                            break;
+                        }
+                    }
+                    if (!nvBuffer)
+                        throw exception("A QueueBuffer request has an invalid NvMap Handle ({}) and ID ({})", surface.nvmapHandle, handle.nvmapId);
                 }
             }
-            if (!nvBuffer)
-                throw exception("A QueueBuffer request has an invalid NVMap Handle ({}) and ID ({})", gbpBuffer.nvmapHandle, gbpBuffer.nvmapId);
+
+            if (surface.size > (nvBuffer->size - surface.offset))
+                throw exception("Surface doesn't fit into NvMap mapping of size 0x{:X} when mapped at 0x{:X} -> 0x{:X}", nvBuffer->size, surface.offset, surface.offset + surface.size);
+
+            gpu::texture::TileMode tileMode;
+            gpu::texture::TileConfig tileConfig{};
+            if (surface.layout == NvSurfaceLayout::Blocklinear) {
+                tileMode = gpu::texture::TileMode::Block;
+                tileConfig = {
+                    .surfaceWidth = static_cast<u16>(surface.width),
+                    .blockHeight = static_cast<u8>(1U << surface.blockHeightLog2),
```

Comment:
blockHeightBits is a much better varname for surface.blockHeightLog2

Reply:
The units involved aren't bits though, height isn't measured in bits and I think `log2` is consistent with the TRM